### PR TITLE
Add inplace abn op

### PR DESCRIPTION
**Pull request by heavengate**
**Add inplace_abn OP**
- original PR: https://github.com/PaddlePaddle/Paddle/pull/19465
- `inplace_abn_grad` only takes `Y` as input, remove input `X`
- add `inplace_abn` Python API instead of implement in `batch_norm` python API to specify in place instead of `build_strategy.emable_inplace` because `batch_norm_grad` takes `X` as input, `inplace_abn_grad` takes `Y` as input, `OpMaker` is different，but `op_type` should be choose in python API，where `enable_inplace` strategy is unknown

中文文档: https://github.com/PaddlePaddle/FluidDoc/pull/1938

屏幕快照 2020-03-23 下午9 39 24
屏幕快照 2020-03-23 下午9 39 34
屏幕快照 2020-03-23 下午9 39 43

### CODE REVIEW

**Review by qingqing01 — DISMISSED** [review]

File: paddle/fluid/operators/batch_norm_op.cc [hunk 1]
```diff
@@ -204,8 +210,13 @@ void BatchNormOpMaker::Make() {
...
       .AddCustomChecker([](const float &epsilon) {
-        PADDLE_ENFORCE(epsilon >= 0.0f && epsilon <= 0.001f,
-                       "'epsilon' should be between 0.0 and 0.001.");
+        PADDLE_ENFORCE_GE(
+            epsilon, 0.0f,
+            platform::errors::InvalidArgument(
+                "'epsilon' should be greater or equal than 0.0."));
+        PADDLE_ENFORCE_LE(epsilon, 0.001f,
+                          platform::errors::InvalidArgument(
+                              "'epsilon' should be less or euqal than 0.001."));
```
Comment (sneaxiy) [inline paddle/fluid/operators/batch_norm_op.cc]: `euqal` -> `equal`?

File: paddle/fluid/operators/batch_norm_op.cc [hunk 2]
```diff
@@ -554,6 +589,20 @@ class BatchNormGradKernel<platform::CPUDeviceContext, T>
...
 
+    // batch_norm with inplace as false will take X as grad input, which
+    // is same as cuDNN batch_norm backward calculation, batch_norm
+    // with inplace as true only take Y as input and X should be calculate
+    // by inverse operation of batch_norm on Y
+    const Tensor *x;
+    bool is_inplace;
+    if (ctx.HasInput("Y")) {
+      x = ctx.Input<Tensor>("Y");
+      is_inplace = true;
```
Comment (sneaxiy) [inline paddle/fluid/operators/batch_norm_op.cc:604]: `PADDLE_ENFORCE(dy == dx)` to avoid bug?

File: paddle/fluid/operators/batch_norm_op.cc [hunk 3]
```diff
@@ -554,6 +589,20 @@ class BatchNormGradKernel<platform::CPUDeviceContext, T>
...
+    } else {
+      x = ctx.Input<Tensor>("X");
+      is_inplace = false;
```
Comment (sneaxiy) [inline paddle/fluid/operators/batch_norm_op.cc:610]: Also, `PADDLE_ENFORCE(dy != dx)`.

File: paddle/fluid/operators/batch_norm_op.cu [hunk 4]
```diff
@@ -355,6 +356,39 @@ static __global__ void KeBNBackwardData(const T *dy,
...
+class InplaceHelper {
+ public:
+  void operator()(const framework::DataLayout layout, T *x,
+                  const BatchNormParamType<T> *scale,
+                  const BatchNormParamType<T> *bias,
+                  const BatchNormParamType<T> *mean,
+                  const BatchNormParamType<T> *variance, double epsilon, int C,
+                  int M, const int num, const T *y, int grid2, const int block,
+                  const cudaStream_t &stream) {
+    KeBNRestoreData<<<grid2, block, 0, stream>>>(
```
Comment (sneaxiy) [inline paddle/fluid/operators/batch_norm_op.cu:389]: Add `PADDLE_ENFORCE(x == y)` ?

File: paddle/fluid/operators/batch_norm_op.cu [hunk 5]
```diff
@@ -417,17 +451,33 @@ class BatchNormGradKernel<platform::CUDADeviceContext, T>
...
     const auto *scale = ctx.Input<Tensor>("Scale");
+    const auto *bias = ctx.Input<Tensor>("Bias");
+
+    // batch_norm with inplace as false will take X as grad input, which
+    // is same as cuDNN batch_norm backward calculation, batch_norm
+    // with inplace as true only take Y as input and X should be calculate
+    // by inverse operation of batch_norm on Y
+    const Tensor *x;
+    bool is_inplace;
+    if (ctx.HasInput("Y")) {
```
Comment (sneaxiy) [inline paddle/fluid/operators/batch_norm_op.cu:479]: Also, enforce `dx == dy`.

File: paddle/fluid/operators/batch_norm_op.cu [hunk 6]
```diff
@@ -417,17 +451,33 @@ class BatchNormGradKernel<platform::CUDADeviceContext, T>
...
+      x = ctx.Input<Tensor>("Y");
+      is_inplace = true;
+    } else {
+      x = ctx.Input<Tensor>("X");
```
Comment (sneaxiy) [inline paddle/fluid/operators/batch_norm_op.cu:486]: Also, enforce `dx != dy`.

File: paddle/fluid/operators/inplace_abn_op.cc [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+  }
+};
+
+class InplaceABNOpMaker : public paddle::operators::BatchNormOpMaker {
+ public:
+  void Make() override {
+    BatchNormOpMaker::Make();
+    AddAttr<std::string>(
+        "activation",
+        "(enum string, default identity, can be identity|elu|leakyrelu) "
```
Comment (sneaxiy) [inline paddle/fluid/operators/inplace_abn_op.cc]: leaky_relu?

File: paddle/fluid/operators/inplace_abn_op.cc [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    op->SetAttrMap(this->Attrs());
+
+    op->SetOutput(framework::GradVarName("X"), this->InputGrad("X"));
+    op->SetOutput(framework::GradVarName("Scale"), this->InputGrad("Scale"));
+    op->SetOutput(framework::GradVarName("Bias"), this->InputGrad("Bias"));
+  }
+};
+
+DECLARE_INPLACE_OP_INFERER(InplaceABNInplaceInferer, {"X", "Y"});
+DECLARE_INPLACE_OP_INFERER(InplaceABNGradInplaceInferer,
```
Comment (sneaxiy) [inline paddle/fluid/operators/inplace_abn_op.cc]: If you add `PADDLE_ENFORCE` inside kernel, please remove this unnecessary configurations of InplaceOpInferer.

File: paddle/fluid/operators/inplace_abn_op.cc [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+                           {framework::GradVarName("Y"),
+                            framework::GradVarName("X")});
+
+template <typename DeviceContext, typename T>
+class InplaceABNKernel
+    : public paddle::operators::BatchNormKernel<DeviceContext, T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* y = ctx.Output<Tensor>("Y");
```
Comment (sneaxiy) [inline paddle/fluid/operators/inplace_abn_op.cc:150]: Enforce `x == y`.

File: paddle/fluid/operators/inplace_abn_op.cc [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+  }
+};
+
+template <typename DeviceContext, typename T>
+class InplaceABNGradKernel
+    : public paddle::operators::BatchNormGradKernel<DeviceContext, T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* y = ctx.Input<Tensor>("Y");
+    auto* d_y = ctx.Input<Tensor>(framework::GradVarName("Y"));
```
Comment (sneaxiy) [inline paddle/fluid/operators/inplace_abn_op.cc:170]: Enforce `dy == dx`.

File: paddle/fluid/operators/inplace_abn_op.cu [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+namespace operators {
+
+template <typename DeviceContext, typename T>
+class InplaceABNKernel
+    : public paddle::operators::SyncBatchNormKernel<DeviceContext, T>,
+      public paddle::operators::BatchNormKernel<DeviceContext, T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* y = ctx.Output<Tensor>("Y");
+    auto activation =
```
Comment (sneaxiy) [inline paddle/fluid/operators/inplace_abn_op.cu:32]: Enforce `x == y`.

File: paddle/fluid/operators/inplace_abn_op.cu [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+// Deriving the Gradient for the Backward Pass of Batch Normalization
+// https://kevinzakka.github.io/2016/09/14/batch_normalization/
+template <typename DeviceContext, typename T>
+class InplaceABNGradKernel
+    : public paddle::operators::SyncBatchNormGradKernel<DeviceContext, T>,
+      public paddle::operators::BatchNormGradKernel<DeviceContext, T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    const auto* y = ctx.Input<Tensor>("Y");
```
Comment (sneaxiy) [inline paddle/fluid/operators/inplace_abn_op.cu:56]: Enforce `dx == dy`.

File: paddle/fluid/operators/batch_norm_op.cc (hunk 3 above)
Comment (heavengate) [reply paddle/fluid/operators/batch_norm_op.cc:610]: Done, thanks!

File: paddle/fluid/operators/batch_norm_op.cc (hunk 2 above)
Comment (heavengate) [reply paddle/fluid/operators/batch_norm_op.cc:604]: Done, thanks!

File: paddle/fluid/operators/batch_norm_op.cc (hunk 3 above)
Comment (heavengate) [reply paddle/fluid/operators/batch_norm_op.cc]: Done, thanks!

File: paddle/fluid/operators/batch_norm_op.cu (hunk 4 above)
Comment (heavengate) [reply paddle/fluid/operators/batch_norm_op.cu:389]: Done, thanks!

File: paddle/fluid/operators/batch_norm_op.cu (hunk 5 above)
Comment (heavengate) [reply paddle/fluid/operators/batch_norm_op.cu:479]: Done, thanks!

File: paddle/fluid/operators/batch_norm_op.cu (hunk 6 above)
Comment (heavengate) [reply paddle/fluid/operators/batch_norm_op.cu:486]: Done, thanks!

File: paddle/fluid/operators/inplace_abn_op.cc (hunk 9 above)
Comment (heavengate) [reply paddle/fluid/operators/inplace_abn_op.cc:150]: Done, thanks!

File: paddle/fluid/operators/inplace_abn_op.cc (hunk 10 above)
Comment (heavengate) [reply paddle/fluid/operators/inplace_abn_op.cc:170]: Done, thanks!

File: paddle/fluid/operators/inplace_abn_op.cu (hunk 11 above)
Comment (heavengate) [reply paddle/fluid/operators/inplace_abn_op.cu:32]: Done, thanks!

File: paddle/fluid/operators/inplace_abn_op.cu (hunk 12 above)
Comment (heavengate) [reply paddle/fluid/operators/inplace_abn_op.cu:56]: Done, thanks!

File: paddle/fluid/operators/batch_norm_op.cc (hunk 1 above)
Comment (heavengate) [reply paddle/fluid/operators/batch_norm_op.cc]: Done, thanks!

File: paddle/fluid/operators/inplace_abn_op.cc (hunk 7 above)
Comment (heavengate) [reply paddle/fluid/operators/inplace_abn_op.cc]: Done, thanks!

File: paddle/fluid/operators/inplace_abn_op.cc (hunk 8 above)
Comment (heavengate) [reply paddle/fluid/operators/inplace_abn_op.cc]: Done, thanks!

**Review by Boyan-Liu — APPROVED** [review]

**Review by qingqing01 — APPROVED** [review]

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Review by raindrops2sea — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**